nestkernel/eprop_archiving_node.cpp

```diff
@@ -190,12 +192,21 @@ nest::EpropArchivingNode::erase_unneeded_update_history()
 void
 nest::EpropArchivingNode::erase_unneeded_eprop_history()
 {
+  if ( eprop_history_.empty()  // nothing to remove
+    or update_history_.empty() // no time markers to check
```

Comment:
```suggestion
  if ( update_history_.empty() // no time markers to check
```

I just checked and the `eprop_history_` guard is not necessary for the tests to run through. `eprop_history_.erase(eprop_history_.end(), eprop_history_.end())` simply leaves the vector unchanged.

Reply:
That is right, but if `eprop_history_` is empty while `update_history_` contains something, we would run through the for loop over the `update_history_` simply to do nothing. To avoid wasting time on the loop for nothing, I added the check on `eprop_history_`. If, by construction, `eprop_history_` must always be empty if `update_history_` is empty, one can drop the check on `eprop_history_`. But then on should add a comment pointing this out.

Reply:
To the first point: `erase_unneeded_update_history` is always called before `erase_unneeded_eprop_history`. Corresponding to the erased values in `update_history` , entries from `eprop_history` are erased in `erase_unneeded_eprop_history`. So, `eprop_history_` cannot be empty if `update_history_` contains something.

To the second point: At the beginning of the function `erase_update_history_`, `eprop_history_` may contain values even if `update_history_` is empty. In this case, the function returns already because `update_history_.empty()` is fulfilled.

This is my understanding, maybe @JesusEV could double-check this.

Reply:
Dependencies between methods are always a bit tricky. I might be that for some reason or another, order suddenly changes (that might even happen through merges in exceptional cases). Then, being defensive does not hurt. And the check for `update_history.empty()` requires negligible time compared to all else in this method. So I would prefer to keep the test in place.

Reply:
The test failures appear to be a consequence of simulating eprop neurons without incorporating eprop synapses. It's the eprop synapses that, through the `check_connection()` method, initiate the `update_history_` buffers. Given that these buffers start off as empty vectors, operations such as long `t = update_history_.begin()->t_;` and similar are likely to be problematic.

I agree about leaning on the safe side and have the guards and checks in place. They will be useful when conducting  benchmarks, where comparing the performance of eprop neurons with eprop synapses and with static synapses will be a step. Without the guards and checks such comparisons would be compromised. 

In that case we have to take other measures to prevent the size of the eprop history from getting out of control (the present function `erase_unneeded_eprop_history()` is the one that cleans up the eprop history and depends on `update_history`, so we need another mechanism to take care of the eprop_history (which by the way is unnecesary in case we have static synapses), but @akorgor already has some good ideas to address this.

Reply:
@JesusEV Yes, the problem was simulations without synapses in place. That should always work for neurons, and now it does again.

I don't think filling up of history would be a problem—the same happens for ArchivingNode if the archive is not emptied by stdp, I think.

In any meaningful benchmark including eprop neurons, they must be connected with eprop synapses, and then history should be cleared, should it not?

Reply:
@heplesser  Yes, cleaning occurs without problems if there are eprop synapses. However, an interesting benchmark is to compare the execution times for a large network with eprop synapses vs. without eprop synapses (i.e., with static synapses), with the idea of ​​getting an notion of ​​the overhead introduced by the learning rule. Of course in that case the learning rate has to be zero to ensure that the dynamics in the two cases are identical.

My concern is letting the eprop_history grow uncontrollably in the case of static synapses, but that may not be a problem after all.